lib/Sema/CSSimplify.cpp

```diff
@@ -3717,6 +3725,13 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
         // trivially solved.
         return getTypeMatchSuccess();
       }
+    } else if (typeVar1) {
+      // Recording the fixed type bindind source locator to this variable
+      // if it was not able to be recorded on attempting type variable
+      // binding, meaning the fixed type can come from a constraint.
+      if (!hasRecordedTypeVariableBindingLocator(typeVar1))
+        recordTypeVariableBindingLocator(typeVar1,
+                                         getConstraintLocator(locator));
```

Comment:
Humm.. so we only record the binding on `assignFixedType` or just if we didn't record it already on `TypeVariableBinding`, I remember when I tried to record only all in `assignFixedType` I had problems because we don't have the source locator there. It works if we pass the source locator as the bind locator like here  9e85a18, but it ends up breaking other diagnostics that rely on the other locator... maybe I didn't do this correctly. I'll get back to this as soon as possible.

Reply:
I think recording should only be done by `assignFixedType` because that's the point which is supposed to be binding type variables. What we need to is provide `Bind` constraint with correct "source" locator which fixed by https://github.com/apple/swift/pull/28752.

Reply:
Thank you @xedin I'm getting back to this, may have something soon :)

Reply:
Great! :)

Reply:
@xedin Just a question, is there any other place where a fixed type could be set for a given type variable in the system besides `assignFixedType`? Because there are some cases I'm seeing where even though a type variable has a fixed type in the system, it wasn't set by `assignFixedType` because there is no locator recorded.

Reply:
Yeah, that's weird ... I think most work is almost done here, the only thing missing is to find out why those edge cases fixed types on closure aren't being set on `assignFixedTypes` :)

Reply:
Those are the false positives we are getting
```swift
func f11295_3<A: Hashable, R>(
  f: @escaping ((A) -> R, A) -> R
) -> ((A) -> R) {
  return { (a: A) -> R in
    let r: R! 
    return r
  }
}

_ = f11295_3 { 
  (f, n) in
  f(n as Int) as Int // OK
}

_ = f11295_3 {
  (f, n) in
  n < 2 ? n as Int : f(n - 1) + f(n - 2) // OK
}

```

The problem is when types are bound via conversion constraints e.g. Conversion, ArgumentConversion, OperatorArgumentConversion. Where the `assignFixedType` is not called to set the fixed to the type variable before where we perform the `RemoveUnnecessaryCoercion::attempt` (maybe is set some other place I couldn't find yet). 
In fact, it is being called only after the check-in `applySolution`. So a possible solution, would be to record this in matchTypes 8340e7516a2218b3944cd771f774fe93a00c4c8b for those constraints. But although this make sense and fixes the issue I'm not sure how correct this would be.

Reply:
@LucianoPAlmeida Do you mean `n` in this case or result of invocation of `f`? I can see with `-debug-constraints` that result of `f(n as Int)` is represented as `$T11` which gets attempted as disjunction choice `(attempting disjunction choice $T11 conv Int [[locator@0x7fb4d281e128 [Coerce@<REPL Input>:3:17]]];`. Same thing happens with `$T8` which represents `n`.

Reply:
Hi @xedin :)
Exactly, in this example, n bounds the type `A` and the result of `f` bounds the type `R`. When debugging this I was tracking down $T8 the more specifically, where I note that it wasn't hitting `assignFixedType` for this typeVar ... so my guess was it was set in another place I couldn't find yet

Reply:
Disjunction choices are bound by `resolveOverload`, AFAIK it should be using `Bind` constraint (which results in `assignFixedType`), maybe there is something interesting going on there...

Reply:
Yeah, I'm investigating that since yesterday but couldn't find what is happening ... I'm getting back to this in a moment, just finishing to squash the commit right now :)

Reply:
@xedin when you mean "resolveOverload should be using Bind constraint (which results in assignFixedType)" could that be that one call in `applySolution` in CSSolver.cpp? One thing indeed is that there is a call to `assignFixedTypes` for this type var, but only after the attempt of the fix (that is being done in matchTypes)...

Reply:
An interesting thing is that when attempting a disjunction choice for an explicit conversion, in `propagateConversionInfo` there for both $T8 and $T11 it is exiting early [here] (https://github.com/apple/swift/blob/439d99cca3ca6503af954eca989a851f0f244a9a/lib/Sema/CSSolver.cpp#L2300) because the binding involves other type variables, so the `Bind` [in lines below](https://github.com/apple/swift/blob/439d99cca3ca6503af954eca989a851f0f244a9a/lib/Sema/CSSolver.cpp#L2320) isn't done. Maybe this is the expected behavior, I really don't know that part to deep, but It was a suspicious thing, I just recorded the binding source here and it solves the issues, but something that is still suspicious is that this fixed type is set somewhere and I can't find where 😆 
I'll be back to this tomorrow to investigate it more :)

Reply:
<img width="440" alt="Screen Shot 2019-12-18 at 00 32 21" src="https://user-images.githubusercontent.com/8292651/71053715-1c959a00-212e-11ea-85c0-f94531ae6cdd.png">

Looking to the stack trace of the check(conditional breakpoint where typeVariable.getID() == 8), I was wondering that it makes sense to not have the assignment of the fixed type performed yet because at this point we are still attempting a type variable binding for the $T8. So, that is why recording the potential binding as source([here](97b7ccccc78b5545334a549ba91f03dc8ac9aab0)) works... because we are recording the potential binding that is being attempted as the source and since is being bound by the coercion... anyways those are assumptions that I have, but since I don't know the inner workings of somethings yet, maybe I'm missing some detail. 
So let me know what you think @xedin and if this makes some sense :))

Reply:
Just to elaborate on my previous comment. I was talking in general - when there is a disjunction choice to attempt it calls into `resolveOverload` which takes a type variable "to bind", represented by `boundType` parameter and a choice to attempt represented by `choice` parameter. Inside it computes the reference type to assign to `boundType` and calls into `bindOverloadType` to assign it. You can see that for regular overloads it adds a `Bind` constraint and uses locator provided by `resolveOverload` which is the locator associated with a particular disjunction choice. `Bind` constraint decays into `assignFixedType`. That's how it works if choice constraint is `BindOverload`. 

If the choice constraint is not `BindOverload` it would do what `simplifyConstraint` tells it to do based on its kind, if I remember correctly - for coercion it would create a `Conversion` or `BridgedConversion` constraint which then used as a source for the binding on the left-hand side type variable. 

I can think of only one way when type variable is not bound via `assignFixedType` - when it's merged with other type variable via `mergeEquivalenceClasses`. In all other cases it should go through `assignFixedType` or it's a bug.

Reply:
I got into `mergeEquivalentClasses` while debugging this, maybe there something there, sorry for all the trouble and questions ... some things are just a little harder for me to understand. Thank you for the elaborated comment :)

Reply:
No problem! Happy to help! :)

Reply:
@xedin I put some debug logging into the fix attempt and `assignFixedType` and here is what happening:
```
 $$(assigning fixed type (@escaping (($T2) -> $T3, $T2) -> $T3) -> (($T2) -> $T3) := $T1)
 $$(assigning fixed type (($T2) -> $T3) := $T12)
 $$(assigning fixed type Int := $T4)
 $$(assigning fixed type ($T2) -> $T3 := $T0)
 $$(assigning fixed type ($T2) -> $T3 := $T5)
 $$(assigning fixed type Int := $T3)
 $$(attempting remove unnecessary coercion fix for := $T11)
 $$(assigning fixed type Int := $T2)
 $$(assigning fixed type Int := $T7)
 $$(attempting remove unnecessary coercion fix for := $T8)
 $$(assigning fixed type () := $T4)
 $$(assigning fixed type (Int) -> Int := $T13)
 $$(assigning fixed type Int := $T3)
 $$(assigning fixed type Int := $T8)
 $$(assigning fixed type Int := $T2)
 $$(assigning fixed type Int := $T7)
 $$(assigning fixed type (Int) -> Int := $T6)
 $$(assigning fixed type (Int) -> Int := $T5)
 $$(assigning fixed type Int := $T4)
 $$(assigning fixed type Int := $T11)
 $$(assigning fixed type @lvalue R? := $T0)
 $$(assigning fixed type @lvalue R := $T0)
 $$(assigning fixed type @lvalue R := $T0)
```
Note that there is in fact an `assignFixedType` call for `$T8`(sometimes `$T10`) but it only happens after the check in match types. I have an proposed approach to fix the issue in b6fa09a01737ebb6f92d40f094a2af830d17f961 
Let me know if this makes sense :)

Reply:
Hm, I'm going to build your branch and look at it under debugger since I still can't wrap my mind around what is going on here :)

Reply:
> Hm, I'm going to build your branch and look at it under debugger since I still can't wrap my mind around what is going on here :)

Thank you so much, really appreciate all your help :)) 

I've also been investigating more too and found some interesting points and tried other things.
What I could understand at the moment is that it seems that there is a fixed type in the system for the `$T8` at the moment of the check, but it is coming from an equivalence class binding $T7 from `savedBindings`. I've been trying an approach of recording the same locator source of `$T7` in `$T8` but it doesn't work because this locator point to the whole `call_expr`. I try that because what I could understand from the code of what equivalent classes are is that it bind type variables to be equivalent in terms of fixed type, meaning that the fixed type bounded to one will be the same for the other. In this case, the fixed type assigned to `$T7` is the same for `$T8` since it is a equivalent class for `$T8`.
Is that correct? Just check if I understand correctly :)

Reply:
Yes, if `$T7` is equivalent to `$T8` it means that both type variables are going to be bound to the same type and have the same flags etc. in each equivalence pair there is representative type variable retrieved by `ConstraintSystem::getRepresentative(TypeVariableType*)`. Representative should always be used to retrieve information about "representee" type variable e.g. non-representative type variables are not used for binding selection and shouldn't be used directly to retrieve recorded locator either. As I mentioned in my previous comment - that's the only situation which makes sense to me when type variable would be missing a direct source location.

Reply:
Humm... thank you for the explanation @xedin :)) 
> As I mentioned in my previous comment - that's the only situation which makes sense to me when the type variable would be missing a direct source location.

Now I see, sorry I miss that, I didn't fully understand that at the time ... have to learn more about it.

So what do you think, should we just ignore and just not warn those cases where we don't have a direct source? 
Or use the current approach b6fa09a where we record the potential binding source(flagged as coming from binding) and use this information on this check? This gives us a potential direct source, and since we are performing a check on a type equals match from a conversion constraint this may be sufficient to assume that this will be the source for the fixed type? 

I think the safer way it is to just ignore and just not warn, although doing this might not warn cases where it probably should be e.g. `var _: (Float) -> Float = { $0 as Float }`... but since they are special cases, it may not be a problem.

What do you think @xedin?

Reply:
I think that check should always use a type variable retuned from `getRepresentative` to retrieve the source of the binding. 

Let me explain how the constraint generation works for the example you mentioned - `var _: (Float) -> Float = { $0 as Float }`. Here `$0` is actually a representation of a parameter type of the closure, to account for `inout (in the function type)` -> `@lvalue (in body)` constraint system would generate a type variable for external use (in the function type of the closure) and another one for use in the body, let's say `$T_ext` (closure type would be `($T_ext) -> $T_result` and `$T_in` in the body. These type variables are going to be connected through `BindParam` constraint so in case `$T_ext` is set to `InOutType` then `$T_in` would become `LValueType`. In most cases we can "contract" that constraint and just merge equivalence classes together (when it's determined that there is no `inout` -> `l-value` translation possible, like you your example where the type of the parameter comes from the context), which means that `$T_ext` is going to be a "representative" for `$T_in`. 

So going back to the check in  b6fa09a - `getRepresentative($T_in)` would produce `$T_out` which is exactly what we need to get the binding source because `$T_out` would be bound to contextual type (because only representative type variables can have potential bindings).

Reply:
> I think that check should always use a type variable returned from `getRepresentative` to retrieve the source of the binding.
> 

Sure, I've tried something similar before (which was in `mergeEquivalentClasses` record the equivalent source as a source for the type variable), and that works fine for this example, but going back to the actual problematic case 
```swift
func f11295_3<A: Hashable, R>(
  f: @escaping ((A) -> R, A) -> R
) -> ((A) -> R) {
  return { (a: A) -> R in
    let r: R! 
    return r
  }
}

_ = f11295_3 { 
  (f, n) in
  f(n as Int) as Int // OK
}
```

Even getting the source locator for the representative of `$T8` in this case `$T7`, the locator source points to a locator that points to a `call_expr` 
```
(call_expr type='<null>' arg_labels=_:
  (declref_expr type='<null>' decl=sr11295.(file).f11295_3 ....
...
...
```
And debugging this the recording of the source seems correct and it is coming from the type variable binding... so since I was not sure if there was a problem there, I got on to try other things :) But now, with your detailed explanation, I think maybe there is another underlying problem there. I'll look it up 

And also, thank you again for the detailed explanation :))

Reply:
No problem! I think the fact that locator points to the call expression is correct because `$T7` is a type variable which represents a "result of the call" and has `FunctionResult` as a last element.

Reply:
Humm ... but the actual locator I'm getting as a source for the representative has the last element as a `TupleElement`. Here is exactly the `typeSourceLocator` path :
```
(lldb) p typeVariable->dump()
(type_variable_type id=8)
(lldb) p representative->dump()
(type_variable_type id=7)
(lldb) p typeSourceLocator->getPath().size()
(size_t) $0 = 4
(lldb) p typeSourceLocator->getPath()[0].getKind()
(swift::constraints::ConstraintLocator::PathElementKind) $1 = ApplyArgument
(lldb) p typeSourceLocator->getPath()[1].getKind()
(swift::constraints::ConstraintLocator::PathElementKind) $2 = ApplyArgToParam
(lldb) p typeSourceLocator->getPath()[2].getKind()
(swift::constraints::ConstraintLocator::PathElementKind) $3 = FunctionArgument
(lldb) p typeSourceLocator->getPath()[3].getKind()
(swift::constraints::ConstraintLocator::PathElementKind) $4 = TupleElement
(lldb) 
```
I really not sure, how to handle this... for now I'm the only this I can think is to skip the warning if locator points to a `TupleElement`...

Reply:
This is the case where we compare argument/parameter in `matchFunctionTypes`. `TupleElement` is used for the comparison but we really should be using a designated locator element similar to `ApplyArgToParam`. We have stubbled on this problem while @hborla was porting tuple conversion mismatches as well. I think Holly is working on improving this, to make it possible to simplify such locators down to anchor.

Reply:
> I think Holly is working on improving this, to make it possible to simplify such locators down to an anchor.

So, by doing that we are going to be able to get to the correct source in this case? There is any other way to get to this info right now without this? 

Also, for this diagnostic in specific, it would be a viable alternative to just skip the warning if `typeSourceLocator` points to a `TupleElement`? Or is it better to wait until we have this right?

Another thing is if that's the only thing left in this PR? Just to know if we are close to getting this in, or if there are maybe other things still to fix here :))

Reply:
@xedin Unfortunately, while revising this, I've just encountered another problem ... :( 
So... I think I'm holding this one for a bit while I try to tackle it in separate in another branch :))

Reply:
@xedin 
The other problem I've mentioned is for this case, kinda similar to others that we had we `CSDiag` and re-typecheck
```swift
class Z0 {}
class TestOverloadSets {
  convenience init(a : Z0) {
    self.init(42 as Int8) // expected-error{{argument labels '(_:)' do not match any available overloads}}
  }
  
  init(value: Int) { /* ... */ }
  init(value: Double) { /* ... */ }
}
```
To emit this diagnostic, it goes into `CSDiag` [here](https://github.com/apple/swift/blob/746b58e8e1c1ef1ac09d8031875fd2a08b65597c/lib/Sema/CSDiag.cpp#L1800) where force re-type checks which produces a false positive diagnostic here. 
I think the check I was doing was not so good, where it was verifying if the expr had already a type applied to it, meaning it was already type-checked and this is a match coming from `CSDiag` to emit diagnostics. I've started an attempt to move this to a fix, [here](https://github.com/LucianoPAlmeida/swift/tree/port-labels-not-match-overloads) but couldn't quite get it correctly because of many issues ... I'll probably need to spend more time on that to get it right. 

But for a solution to this specific issue, I just end up moving the diagnostic for before the re-check since it doesn't necessarily need this, so we diagnose and exit early avoiding the force re-typecheck and we could get rid of the wrong verification on `attempt` fix in 8dc35ead40be3e4c5dac54dede2de3bd5d3bcfcb. 

Sorry for the long text, I just wanted to give a context on the problem that we have right now :))

Reply:
That's something I was worried about happening because there is no logic to detect whether producing warnings is appropriate or not.

I think we have two options here: a. Try to add some special handling to `applySolutionWithFixes` which would detect re-typecheck and skip any warnings generated by the solver (I think it should be possible to archive by checking for presence of `SubExpressionDiagnostics` flag in constraint system options); b. Postpone these changes until after CSDiag is gone;

Regarding question about locator - yes, once there is a special locator for argument/parameter comparison between function types it would be possible to simplify it down to a correct anchor instead of a top-level call expression.

Reply:
Thank you for the answers @xedin 
> I think we have two options here: a. Try to add some special handling to `applySolutionWithFixes` which would detect re-typecheck and skip any warnings generated by the solver (I think it should be possible to archive by checking for presence of `SubExpressionDiagnostics` flag in constraint system options); b. Postpone these changes until after CSDiag is gone;
> 

Nice, I didn't know about this flag, seems to work fine now :))

> Regarding question about locator - yes, once there is a special locator for argument/parameter comparison between function types it would be possible to simplify it down to a correct anchor instead of a top-level call expression.
>

That's nice \o/
You mention @hborla was working on that, is this in progress or scheduled? If not, I'll be happy to try to work on that, maybe is a bit more advanced to me right now and probably will take more time since I don't know exactly how to do this, but I think with some guidance I may be able to tackle it :)) 
So let me know If can do something to move this forward 👍

Reply:
> I think we have two options here: a. Try to add some special handling to `applySolutionWithFixes` which would detect re-typecheck and skip any warnings generated by the solver (I think it should be possible to archive by checking for presence of `SubExpressionDiagnostics` flag in constraint system options); b. Postpone these changes until after CSDiag is gone;
> 

Ahh @xedin, also ... I just check for it on `attempt` for this fix specifically, not in `applySolutionFixes` because I didn't know how it would impact the tests(testing now) and if it breaks something I didn't want to fix all in this PR, so I check specific... but I can do it in another PR and fix all there. What do you think?

Reply:
@xedin Just finish rebasing/updating this with master :)

Reply:
Are you trying to say that this is ready for review again? :) FYI, there is a refresh icon on the side of each reviewer which allows to re-request the review after the changes have been made.

Reply:
Meanwhile, if you are looking for something to work on, there is `FailureDiagnosis:: visitObjectLiteralExpr` which is waiting to be ported :)

Reply:
> Are you trying to say that this is ready for review again? :) FYI, there is a refresh icon on the side of each reviewer which allows to re-request the review after the changes have been made.
> 


Humm... I think the implementation itself is ready for review, but we stuck on the issue with the `TupleElement` locator path. Maybe it could be reviewed and have some points to improve, but since it takes time, let's wait until we have this issue solved and it is ready to go. 

I just wanted to ping you, and meant that since I updated it with master, the only thing left to get this ready is to try to solve the problem and I wanted to know if maybe there is something I can help working on to fix it so we can move this forward :)

> Meanwhile, if you are looking for something to work on, there is FailureDiagnosis:: visitObjectLiteralExpr which is waiting to be ported :)
> 

Awesome, I can take it meanwhile :)) Thank you!